index.js

```diff
@@ -1,38 +1,97 @@
 'use strict'
 
 const fp = require('fastify-plugin')
-const mysql = require('mysql')
 
 function fastifyMysql (fastify, options, next) {
   const name = options.name
   delete options.name
 
-  const pool = mysql.createPool(options.connectionString || options)
-  const db = {
-    connect: onConnect => pool.getConnection(onConnect),
-    pool: pool,
-    query: pool.query.bind(pool)
-  }
+  const connectionType = options.type
+  const usePromise = options.promise
 
-  if (name) {
-    if (!fastify.mysql) {
-      fastify.decorate('mysql', {})
-    }
+  _createConnection(options, (err, db) => {
+    if (err) return next(err)
 
-    fastify.mysql[name] = db
-  } else {
-    if (fastify.mysql) {
-      next(new Error('fastify-mysql has already registered'))
+    const client = (connectionType !== 'connection') ? db.pool : db.connection
+
+    if (name) {
+      if (!fastify.mysql) {
+        fastify.decorate('mysql', {})
+      }
+      if (fastify.mysql[name]) {
+        next(new Error('fastify.mysql.' + name + 'has already registered'))
+        return
+      }
+      fastify.mysql[name] = db
     } else {
-      fastify.mysql = db
+      if (fastify.mysql) {
+        next(new Error('fastify-mysql has already registered'))
+        return
+      } else {
+        fastify.mysql = db
+      }
     }
-  }
 
-  fastify.addHook('onClose', (fastify, done) => pool.end(done))
-  next()
+    if (usePromise) {
+      fastify.addHook('onClose', (fastify, done) => client.end())
```

Comment:
can you remove the `done`  argument?

Reply:
@mcollina  sure, one last thing, should we keep the `connect` util function ? or switch to `getConnection` ?